### PR TITLE
Upgraded to Vert.x 4.2.1

### PR DESCRIPTION
**Pull request by ZhengXinCN**

### CODE REVIEW

**Review by hiddenalpha — APPROVED** [review]

File: src/main/java/org/swisspush/redisques/handler/GetAllLocksHandler.java [hunk 1]
```diff
@@ -27,22 +28,26 @@ public GetAllLocksHandler(Message<JsonObject> event, Optional<Pattern> filterPat
...
-                    if(pattern.matcher(lock).find()){
+                    String lock = locks.get(i).toString();
+                    if (pattern.matcher(lock).find()) {
                         filteredLocks.add(lock);
                     }
                 }
                 result.put("locks", filteredLocks);
             } else {
-                result.put("locks", locks);
+                JsonArray values = new JsonArray();
```
[the remark this answers is not in the document]
Comment (ZhengXinCN) [reply src/main/java/org/swisspush/redisques/handler/GetAllLocksHandler.java:46]: ok

File: src/main/java/org/swisspush/redisques/RedisQues.java [hunk 2]
```diff
@@ -680,32 +691,50 @@ private void deleteAllLocks(Message<JsonObject> event) {
...
+                event.reply(createOkReply().put(VALUE, delManyResult.result().toLong()));
             } else {
                 log.warn("failed to delete locks. Message: " + delManyResult.cause().getMessage());
                 event.reply(createErrorReply().put(MESSAGE, delManyResult.cause().getMessage()));
             }
         });
     }
 
-    private boolean jsonArrayContainsStringsOnly(JsonArray array){
+    //TODO: useless in vertx 4
```
[the remark this answers is not in the document]
Comment (ZhengXinCN) [reply src/main/java/org/swisspush/redisques/RedisQues.java:722]: I keep the function and remove TODO, because this case function still works, may needed in some case

File: src/main/java/org/swisspush/redisques/RedisQues.java [hunk 3]
```diff
@@ -878,19 +907,23 @@ private void resetConsumers() {
...
             }
-            JsonArray keys = keysResult.result();
-            if (keys == null || keys.isEmpty()) {
+            Response keys = keysResult.result();
+            if (keys == null || keys.size() == 0) {
                 log.debug("No consumers found to reset");
                 return;
             }
-            redisClient.delMany(keys.getList(), delManyResult -> {
+            List<String> args = new ArrayList<>();
```
[the remark this answers is not in the document]
Comment (ZhengXinCN) [reply src/main/java/org/swisspush/redisques/RedisQues.java:920]: ok

File: src/main/java/org/swisspush/redisques/RedisQues.java [hunk 4]
```diff
@@ -181,13 +187,11 @@ public void start() {
...
-        this.redisClient = RedisClient.create(vertx, new RedisOptions()
-                .setHost(redisHost)
-                .setPort(redisPort)
-                .setAuth(redisAuth)
-                .setEncoding(redisEncoding));
-
-        this.luaScriptManager = new LuaScriptManager(redisClient);
+        this.redisClient = new RedisClient(vertx, new RedisOptions().setConnectionString("redis://" + redisHost + ":" + redisPort).setPassword(redisAuth));
+        //TODO: vertx4
+        //       .setEncoding(redisEncoding));
```
[the remark this answers is not in the document]
Comment (ZhengXinCN) [reply src/main/java/org/swisspush/redisques/RedisQues.java:192]: I look around there no way to set encoding, that why I put a TODO here.

[the remark this answers is not in the document]
Comment (dominik-cnx) [reply src/main/java/org/swisspush/redisques/RedisQues.java:192]: It was set to

this.redisEncoding = "UTF-8";

before, let's find out to what we get now.

**Review by hiddenalpha — APPROVED** [review]

File: src/main/java/org/swisspush/redisques/RedisQues.java (hunk 4 above)
[the remark this answers is not in the document]
Comment (ZhengXinCN) [reply src/main/java/org/swisspush/redisques/RedisQues.java:192]: read here, new version doesn't serialize binary data any more. 
https://groups.google.com/g/vertx-dev/c/xBElx1yu3pg

[the remark this answers is not in the document]
Comment (dominik-cnx) [reply src/main/java/org/swisspush/redisques/RedisQues.java]: OK, I think it will be fine. Let's keep in mind that we should verify handling of binary data in gateleen once we are through (if not already integration tested).

**Review by dominik-cnx — APPROVED** [review]

**Review by ljucam — APPROVED** [review]

File: src/main/java/org/swisspush/redisques/util/RedisUtils.java [hunk 5]
```diff
@@ -0,0 +1,59 @@
...
+import io.vertx.core.json.JsonArray;
+import io.vertx.core.json.JsonObject;
+
+import java.util.*;
+import java.util.stream.Stream;
+
+/**
+ * Useful utilities for Redis
+ *
+ * @author https://github.com/mcweba [Marc-Andre Weber]
```
Comment (mcweba) [inline src/main/java/org/swisspush/redisques/util/RedisUtils.java:12]: I am not the author in this case ;-)

File: src/test/java/org/swisspush/redisques/RedisQuesTest.java [hunk 6]
```diff
@@ -221,6 +221,7 @@ public void getQueues(TestContext context) {
                 context.assertEquals(OK, message.result().body().getString(STATUS));
                 asyncEnqueue.countDown();
             });
+            Thread.sleep(10);
```
Comment (mcweba) [inline src/test/java/org/swisspush/redisques/RedisQuesTest.java:224]: Why is this necessary? The test will wait until the countDown has gone from 100 to 0

File: src/test/java/org/swisspush/redisques/RedisQuesTest.java [hunk 7]
```diff
@@ -247,6 +248,7 @@ public void getQueuesFiltered(TestContext context) {
...
```
Comment (mcweba) [inline src/test/java/org/swisspush/redisques/RedisQuesTest.java:251]: Why is this necessary? The test will wait until the countDown has gone from 20 to 0

File: src/test/java/org/swisspush/redisques/RedisQuesTest.java [hunk 8]
```diff
@@ -278,6 +280,7 @@ public void getQueuesFilteredInvalidPattern(TestContext context) {
...
```
Comment (mcweba) [inline src/test/java/org/swisspush/redisques/RedisQuesTest.java:283]: Why is this necessary? The test will wait until the countDown has gone from 20 to 0

File: src/test/java/org/swisspush/redisques/RedisQuesTest.java [hunk 9]
```diff
@@ -301,6 +304,7 @@ public void getQueuesCount(TestContext context) {
...
```
Comment (mcweba) [inline src/test/java/org/swisspush/redisques/RedisQuesTest.java:307]: Why is this necessary? The test will wait until the countDown has gone from 100 to 0

File: src/test/java/org/swisspush/redisques/RedisQuesTest.java [hunk 10]
```diff
@@ -323,6 +327,7 @@ public void getQueuesCountFiltered(TestContext context) {
...
```
Comment (mcweba) [inline src/test/java/org/swisspush/redisques/RedisQuesTest.java:330]: Why is this necessary? The test will wait until the countDown has gone from 50 to 0

File: src/test/java/org/swisspush/redisques/RedisQuesTest.java [hunk 11]
```diff
@@ -349,6 +354,7 @@ public void getQueuesCountFilteredInvalidPattern(TestContext context) {
...
```
Comment (mcweba) [inline src/test/java/org/swisspush/redisques/RedisQuesTest.java:357]: Why is this necessary? The test will wait until the countDown has gone from 50 to 0

**Review by mcweba — CHANGES_REQUESTED** [review]

File: src/main/java/org/swisspush/redisques/RedisQues.java [hunk 12]
```diff
@@ -121,27 +122,32 @@ private void handleRegistrationRequest(Message<String> msg) {
             log.debug("RedisQues Got registration request for queue " + queueName + " from consumer: " + uid);
         }
         // Try to register for this queue
-        SetOptions setOptions = new SetOptions().setNX(true).setEX(consumerLockTime);
-        redisClient.setWithOptions(consumersPrefix + queueName, uid, setOptions, event -> {
+        redisAPI.setnx(consumersPrefix + queueName, uid, event -> {
```
[the remark this answers is not in the document]
Comment (dominik-cnx) [reply src/main/java/org/swisspush/redisques/RedisQues.java:125]: Thx @mcweba, I think you have a point here, we need to make it atomic.

Can you confirm that the basic working principle is
---
Add (set) it if not already there, otherwise don't touch (e.g. do not update the expiry).
---
And we do need it to be exactly like this, correct?

File: src/test/java/org/swisspush/redisques/RedisQuesTest.java (hunk 6 above)
Comment (dominik-cnx) [reply src/test/java/org/swisspush/redisques/RedisQuesTest.java:224]: We exhaust the connection pool size without

io.vertx.core.http.ConnectionPoolTooBusyException: Connection pool reached max wait queue size of 24

Maybe we can increase it and then we can remove this again?

File: src/main/java/org/swisspush/redisques/RedisQues.java (hunk 12 above)
[the remark this answers is not in the document]
Comment (mcweba) [reply src/main/java/org/swisspush/redisques/RedisQues.java:125]: @dominik-cnx Yes exactly. It is crucial that an existing entry (of another consumer of this queue) is not overwritten.

File: src/test/java/org/swisspush/redisques/RedisQuesTest.java (hunk 6 above)
Comment (Kusig) [reply src/test/java/org/swisspush/redisques/RedisQuesTest.java:224]: Is this a general new behaviour of vert.x 4? Or even a general new problem which we have to care about in general? 
Sending 

Maybe sending one after the other (in the event handler) might help?

**Review by Kusig — APPROVED** [review]

File: src/test/java/org/swisspush/redisques/RedisQuesTest.java (hunk 6 above)
Comment (dominik-cnx) [reply src/test/java/org/swisspush/redisques/RedisQuesTest.java:224]: I think we should make the connection pool size configurable and figure out what the limit was before (and then maybe use this limit).

Comment (dominik-cnx) [reply src/test/java/org/swisspush/redisques/RedisQuesTest.java]: The commands went through a queue in the past. So we need to keep in mind that the implementation has changed in this regard (we can now exhaust the max number of connections whereas in the past commands where queued)

File: src/main/java/org/swisspush/redisques/RedisQues.java (hunk 12 above)
[the remark this answers is not in the document]
Comment (dominik-cnx) [reply src/main/java/org/swisspush/redisques/RedisQues.java]: @mcweba kindly review the revised implementation, thx.

**Review by Kusig — APPROVED** [review]

**Review by mcweba — APPROVED** [review]
